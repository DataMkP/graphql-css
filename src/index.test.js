/* eslint-env jest */

import React from 'react'
import gql from "graphql-tag";
import {render, Simulate} from 'react-testing-library';
import serializer from 'jest-glamor-react';
import {Div} from 'glamorous';

import gqlCSS, { GqlCSS, GqlCSSProvider, WithGqlCSS, withGqlCSS } from "./index";
import styles from '../examples/styleguide';
import { h1Styles, h2Styles, customH1Styles, stateStyles } from "../examples/styleQueries";

expect.addSnapshotSerializer(serializer)

describe('gqlCSS', () => {
    it('it allows the extraction of styles by passing false', () => {
        const extractedStyles = gqlCSS(styles)(h2Styles, false);
        expect(extractedStyles).toEqual({
            'color': 'green',
            'fontSize': '36px',
            'fontWeight': 700,
            'marginLeft': 32,
          })
    });
  
    it('it returns a styled component by passing a div', () => {
        const StyledComponent = gqlCSS(styles)(h2Styles, 'div');
        const {container} = render(<StyledComponent />)
        expect(container).toMatchSnapshot()
    });

    it('it supports variables and stateful components', () => {
        class StatefulComponent extends React.Component {
            constructor(props) {
                super(props);
                this.state = {
                    variant: "normal",
                };
                this.toggleVariant = this.toggleVariant.bind(this);
            }
        
            toggleVariant() {
                this.setState(state => ({
                    variant: state.variant === "normal" ? "done" : "normal",
                }));
            }
        
            render() {
                const { variant } = this.state;
                const OtherComponent = gqlCSS(styles)(stateStyles, null, { variant });
        
                return (
                    <React.Fragment>
                        <GqlCSS
                            styles={styles}
                            query={stateStyles}
                            variables={{ variant }}
                            onClick={this.toggleVariant}
                            data-testid="stateful-component"
                        >
                            Using stateful component
                        </GqlCSS>
                        <OtherComponent>Other sharing state</OtherComponent>
                    </React.Fragment>
                );
            }
        }

        const {container, queryByTestId} = render(<StatefulComponent />)

        expect(container).toMatchSnapshot()

        Simulate.click(queryByTestId('stateful-component'))

        expect(container).toMatchSnapshot()
    });
});

describe('GqlCSSProvider', () => {
    it('it supports the Provider/Subscriber pattern', () => {
        const SubscriberComponent = () => (
            <div>
                <div>
                    <span>
                        <GqlCSS query={h2Styles}>Using provider</GqlCSS>
                    </span>
                </div>
                <GqlCSS query={h1Styles} css={{ marginTop: 30 }}>
                    Also using provider
                </GqlCSS>
            </div>
        );
        const {container} = render(<GqlCSSProvider styles={styles}><SubscriberComponent/></GqlCSSProvider>)

        expect(container).toMatchSnapshot()
    });
});

describe('WithGqlCSS', () => {
    it('it supports the render props pattern', () => {
        const {container} = render(
            <WithGqlCSS styles={styles} query={h2Styles}>
                {({ gqlStyles }) => <div style={gqlStyles}>Render props component</div>}
            </WithGqlCSS>
        );

        expect(container).toMatchSnapshot()
    });
});

describe('withGqlCSS', () => {
    it('it supports the HOC pattern', () => {
        const ExistingComponent = ({ gqlStyles, ...rest }) => (
            <Div css={gqlStyles} {...rest}>test</Div>
        );

        const HOC = withGqlCSS(styles, customH1Styles)(ExistingComponent)
        const {container} = render(<HOC />)

        expect(container).toMatchSnapshot()
    });
});