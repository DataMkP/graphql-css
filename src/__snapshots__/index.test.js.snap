// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GqlCSSProvider it supports the Provider/Subscriber pattern 1`] = `
.css-0,
[data-css-0] {
  font-size: 36px;
  font-weight: 700;
  margin-left: 32px;
  color: green;
}

.css-1,
[data-css-1] {
  font-size: 52px;
  font-weight: 700;
  margin-left: 24px;
  color: red;
  margin-top: 30px;
}

<div>
  <span>
    <div>
      <div>
        <span>
          <div
            class="css-0"
          >
            Using provider
          </div>
        </span>
      </div>
      <div
        class="css-1"
      >
        Also using provider
      </div>
    </div>
  </span>
</div>
`;

exports[`WithGqlCSS it supports the render props pattern 1`] = `
<div>
  <div
    style="font-size: 36px; font-weight: 700; margin-left: 32px; color: green;"
  >
    Render props component
  </div>
</div>
`;

exports[`gqlCSS it returns a styled component by passing a div 1`] = `
.css-0,
[data-css-0] {
  font-size: 36px;
  font-weight: 700;
  margin-left: 32px;
  color: green;
}

<div>
  <div
    class="css-0"
  />
</div>
`;

exports[`gqlCSS it supports variables and stateful components 1`] = `
.css-0,
[data-css-0] {
  font-size: 36px;
  background-color: red;
  padding: 24px;
  cursor: pointer;
}

<div>
  <div
    class="css-0"
    data-testid="stateful-component"
  >
    Using stateful component
  </div>
  <div
    class="css-0"
  >
    Other sharing state
  </div>
</div>
`;

exports[`gqlCSS it supports variables and stateful components 2`] = `
.css-0,
[data-css-0] {
  font-size: 36px;
  background-color: green;
  padding: 24px;
  cursor: pointer;
}

<div>
  <div
    class="css-0"
    data-testid="stateful-component"
  >
    Using stateful component
  </div>
  <div
    class="css-0"
  >
    Other sharing state
  </div>
</div>
`;

exports[`withGqlCSS it supports the HOC pattern 1`] = `
.css-0,
[data-css-0] {
  font-size: 52px;
  font-weight: 700;
  margin-left: 4em;
  color: blue;
}

<div>
  <div
    class="css-0"
  >
    test
  </div>
</div>
`;
